adc: Convert mcux adc16 driver to the new adc api

MaureenHelm opened the pull request:
Converts the mcux adc16 driver to the new adc api and updates tests accordingly.

Depends on #7691

anangl committed:
drivers: adc: Introduce reworked API

This commit replaces the API for ADC drivers with a reworked one.
As requested in the issue #3980, the adc_enable/adc_disable functions
are removed. Additionaly, some new features are introduced, like:
- asynchronous calls
- configuration of channels
- multi-channel sampling

Common parts of code that are supposed to appear in each implementation
of the driver (like locking, synchronization, triggering of consecutive
samplings) are provided in the "adc_context.h" file to keep consistency
with the SPI driver. Syscalls are no longer present in the API because
the functions starting read requests cannot use them, since they can be
provided with a callback that is executed in the ISR context, and there
is no point in supporting syscalls only for the channels configuration.

"adc_api" test is updated and extended with additional test cases,
with intention to show how the API is supposed to be used.
"adc_simple" test is removed as it does not seem to add much value.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: adc: Add shims for nrfx ADC and SAADC drivers

This commit adds translation layers to make nrfx drivers for the nRF
ADC (nRF51 series) and SAADC (nRF52 series) peripherals accessible via
the Zephyr's API. The SAADC peripheral is accessed using nrfx HAL only
as it turns out that usage of the nrfx driver in this case would be
inconvenient and would unnecessarily complicate the shim.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nrf: Add ADC nodes and bindings for nRF SoCs

This commit adds ADC nodes to DTS files for nRF SoCs and introduces
corresponding  bindings for these nodes.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf: Enable ADC nodes in DTS for nRF boards

ADC nodes are enabled in DTS for the following nRF development kits:
- nrf51_pca10028
- nrf52_pca10040
- nrf52_pca20020
- nrf52810_pca10040
- nrf52840_pca10056

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: adc_api: Add configurations for nRF5 boards

This adds to the adc_api test board-specific configurations
for the following boards:
- nrf51_pca10028
- nrf52_pca10040
- nrf52840_pca10056

"adc" is indicated as supported in the yaml files for these boards
so that they are included in sanitycheck.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: adc: Select HAS_DTS_ADC for all ADC drivers

To require relevant DTS entries on all platforms providing the ADC
driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

MaureenHelm committed:
adc: Convert mcux adc16 driver to the new adc api

Major rework of the mcux adc16 driver to convert it to the new adc api.
Currently supports a subset of the api features including synchronous
and asynchronous reads, and consecutive reads triggered by the kernel
timer.

Does not yet support some of the channel configuration options such as
gain and reference voltage because the hardware only allows these
options to be configured by peripheral instance rather than by channel.
The values are currently hardcoded in the driver, but in the future we
could introduce some flexibility per instance via device tree
attributes.

Does not yet support consecutive reads triggered by a hardware timer.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: adc_api: Add support for kinetis boards

Adds board-specific macros to enable the adc_api test to run on kinetis
boards. This restores the test to support all the nxp boards that were
supported before the new adc api was introduced.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

anangl commented:
@MaureenHelm My original intention was to make channels within API independent of any physical ADC inputs (now I know it was not necessarily a good idea). That's why in the test I used fixed values for channel identifiers (0 and 2 because I just wanted to check that non-continuous channel identifiers are also handled correctly).
I updated my PR, so after rebasing, your second commit should be no longer needed. You'll need to adjust a bit the third one. I hope it won't be a big problem.

anangl commented on drivers/adc/adc_mcux_adc16.c:
```diff
@@ -1,69 +1,171 @@
+	if (sequence->resolution != 12) {
+		SYS_LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
```
You should also check if any channels were selected and if only the valid ones.

MaureenHelm commented on drivers/adc/adc_mcux_adc16.c:
```diff
@@ -1,69 +1,171 @@
 
-		k_sem_take(&data->sync, K_FOREVER);
+	if (ctx->sampling_index == 0) {
+		data->buffer = ctx->sequence->buffer;
```
Good catch, thank you. The improved test did make it fail, and now it passes after fixing up the buffer pointer initialization/reinitialization.

MaureenHelm commented:
> I updated my PR, so after rebasing, your second commit should be no longer needed. You'll need to adjust a bit the third one. I hope it won't be a big problem.

Not a problem. I've rebased and removed the now-obsolete commit.
